File: Advanced_Robotics_Course/Chapter16/Lesson3/Chapter16_Lesson3.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <set>
#include <utility>

struct State2D {
    int i, j;
    bool operator<(const State2D& other) const {
        if (i != other.i) return i < other.i;
        return j < other.j;
    }
    bool operator==(const State2D& other) const {
        return i == other.i && j == other.j;
    }
};

struct TS {
    std::set<State2D> states;
    State2D initial;
    std::vector<std::string> actions{"up", "down", "left", "right"};
    std::map<std::pair<State2D,std::string>, std::set<State2D> > trans;
    std::map<State2D, std::set<std::string> > label;
};

struct Buchi {
    // F goal
    std::set<std::string> states{"q0", "q1"};
    std::string initial{"q0"};
    std::set<std::string> accepting{"q1"};

    std::string delta(const std::string& q,
                      const std::set<std::string>& sigma) const {
        bool has_goal = sigma.find("goal") != sigma.end();
        if (q == "q0") {
            return has_goal ? "q1" : "q0";
        } else {
            return "q1";
        }
    }
};

using ProdState = std::pair<State2D,std::string>;

struct ProdStateLess {
    bool operator()(const ProdState& a, const ProdState& b) const {
        if (a.first < b.first) return true;
        if (b.first < a.first) return false;
        return a.second < b.second;
    }
};

std::vector<std::pair<ProdState,std::string> >
productSuccessors(const TS& ts, const Buchi& ba,
                  const ProdState& ps) {
    std::vector<std::pair<ProdState,std::string> > out;
    const State2D& s = ps.first;
    const std::string& q = ps.second;
    auto it_label = ts.label.find(s);
    const std::set<std::string>& sigma =
        (it_label != ts.label.end()) ? it_label->second
                                     : *new std::set<std::string>();
    for (const auto& act : ts.actions) {
        auto it = ts.trans.find(std::make_pair(s, act));
        if (it == ts.trans.end()) continue;
        for (const auto& s_next : it->second) {
            std::string q_next = ba.delta(q, sigma);
            out.push_back({ProdState{s_next, q_next}, act});
        }
    }
    return out;
}

int main() {
    TS ts;
    // Here you would fill ts.states, ts.initial, ts.trans, ts.label

    Buchi ba;
    ProdState start{ts.initial, ba.initial};

    std::queue<ProdState> q;
    std::set<ProdState,ProdStateLess> visited;
    q.push(start);
    visited.insert(start);

    while (!q.empty()) {
        ProdState cur = q.front(); q.pop();
        const std::string& q_ba = cur.second;
        bool is_accepting = ba.accepting.find(q_ba) != ba.accepting.end();
        if (is_accepting) {
            std::cout << "Found reachable accepting product state" << std::endl;
            break;
        }
        for (auto& succ : productSuccessors(ts, ba, cur)) {
            const ProdState& nxt = succ.first;
            if (visited.find(nxt) == visited.end()) {
                visited.insert(nxt);
                q.push(nxt);
            }
        }
    }
    return 0;
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter16/Lesson3/Chapter16_Lesson3.py
from collections import defaultdict, deque
from typing import Dict, Set, Tuple, List

State = Tuple[int, int]  # grid coordinates (i, j)

class TransitionSystem:
    def __init__(self):
        self.states: Set[State] = set()
        self.initial: State = None
        self.actions = ["up", "down", "left", "right"]
        self.transitions: Dict[Tuple[State, str], Set[State]] = defaultdict(set)
        self.ap: Set[str] = {"goal", "unsafe"}
        self.label: Dict[State, Set[str]] = dict()

    def add_state(self, s: State, is_initial: bool = False,
                  props: Set[str] = None):
        self.states.add(s)
        if is_initial:
            self.initial = s
        self.label[s] = set(props) if props is not None else set()

    def add_transition(self, s: State, a: str, s_next: State):
        self.transitions[(s, a)].add(s_next)

class BuchiAutomaton:
    def __init__(self):
        # For phi = F goal
        self.states: Set[str] = {"q0", "q1"}
        self.initial: str = "q0"
        self.accepting: Set[str] = {"q1"}

    def delta(self, q: str, sigma: Set[str]) -> str:
        has_goal = "goal" in sigma
        if q == "q0":
            return "q1" if has_goal else "q0"
        else:
            # q1 is absorbing
            return "q1"

def build_grid_ts(width: int, height: int,
                  obstacles: Set[State],
                  goal: State,
                  start: State) -> TransitionSystem:
    ts = TransitionSystem()
    for i in range(height):
        for j in range(width):
            s = (i, j)
            if s in obstacles:
                continue
            props = set()
            if s == goal:
                props.add("goal")
            ts.add_state(s, is_initial=(s == start), props=props)

    # 4-neighborhood
    dirs = {"up": (-1, 0), "down": (1, 0),
            "left": (0, -1), "right": (0, 1)}
    for s in ts.states:
        i, j = s
        for a, (di, dj) in dirs.items():
            s_next = (i + di, j + dj)
            if s_next in ts.states:
                ts.add_transition(s, a, s_next)
    return ts

def product_initial(ts: TransitionSystem, ba: BuchiAutomaton):
    return (ts.initial, ba.initial)

def product_successors(ts: TransitionSystem, ba: BuchiAutomaton,
                       s_prod: Tuple[State, str]) -> List[Tuple[Tuple[State, str], str]]:
    (s, q) = s_prod
    result = []
    for a in ts.actions:
        for s_next in ts.transitions.get((s, a), []):
            sigma = ts.label[s]
            q_next = ba.delta(q, sigma)
            result.append(((s_next, q_next), a))
    return result

def find_accepting_lasso(ts: TransitionSystem, ba: BuchiAutomaton):
    start = product_initial(ts, ba)
    # BFS for reachable states and predecessor map
    queue = deque([start])
    pred: Dict[Tuple[State, str], Tuple[Tuple[State, str], str]] = {}
    visited = {start}
    accepting_states = set()

    while queue:
        cur = queue.popleft()
        (s, q) = cur
        if q in ba.accepting:
            accepting_states.add(cur)
        for nxt, act in product_successors(ts, ba, cur):
            if nxt not in visited:
                visited.add(nxt)
                pred[nxt] = (cur, act)
                queue.append(nxt)

    if not accepting_states:
        return None  # no plan

    # For simplicity, pick one accepting state and search for a cycle to itself
    acc = next(iter(accepting_states))

    # DFS within visited states to find a cycle starting and ending at acc
    def dfs_cycle(cur, target, visited_local, path):
        if cur == target and path:
            return path
        visited_local.add(cur)
        for nxt, act in product_successors(ts, ba, cur):
            if nxt not in visited or nxt not in visited_local:
                res = dfs_cycle(nxt, target, visited_local, path + [(nxt, act)])
                if res is not None:
                    return res
        return None

    # Recover prefix to acc
    prefix_states_actions = []
    cur = acc
    while cur != start:
        prev, act = pred[cur]
        prefix_states_actions.append((cur, act))
        cur = prev
    prefix_states_actions.reverse()

    cycle = dfs_cycle(acc, acc, set(), [])
    if cycle is None:
        return None  # no lasso found

    return start, prefix_states_actions, cycle

if __name__ == "__main__":
    width, height = 4, 3
    obstacles = {(1, 1)}
    goal = (0, 3)
    start = (2, 0)
    ts = build_grid_ts(width, height, obstacles, goal, start)
    ba = BuchiAutomaton()
    lasso = find_accepting_lasso(ts, ba)
    if lasso is None:
        print("No plan satisfying F(goal) found.")
    else:
        start, prefix, cycle = lasso
        print("Prefix:")
        print(prefix)
        print("Cycle:")
        print(cycle)
      
